### ISSUE TITLE
Regarding MacBook OS update

### ISSUE THREAD

**Issue by thgr8ganzi** [rodyager/RWTS-PDFwriter#50]
The program does not work after updating my MacBook. Please confirm.

It seems that the program is not recognized as a virtual printer.

**Comment by rodyager** [comment]
I am not sure what is happening on your MacBook. PDFWriter works as expected on my MacBook running MacOS 15 (Sequoia) and also worked on MacOS 14 (Sonoma)

**Comment by thgr8ganzi** [comment]
Yes, thank you. I will try again and let you know if there are any problems.

**Comment by r3garciaz** [comment]
how can i print in Sonoma? I have a wine app that don't print. Always say "Incoming"


<img width="896" alt="Screenshot 2024-11-11 at 18 17 09" src="https://github.com/user-attachments/assets/7d7759d7-bdfc-4be8-95ab-22ea46502761">

**Comment by rodyager** [comment]
You can’t from wine.  Wine draws its output in postscript and the Sonoma (and later) MacOS does not support postscript to pdf conversion in its CUPS processing chain.


Rod

On 12 Nov 2024, at 8:20 am, Reinaldo García Z ***@***.***> wrote:



how can i print in Sonoma? I have a wine app that don't print. Always say "Incoming"

Screenshot.2024-11-11.at.18.17.09.png (view on web)<https://github.com/user-attachments/assets/7d7759d7-bdfc-4be8-95ab-22ea46502761>

**Comment by r3garciaz** [comment]
Ok, I get. Thanks for your reply 

Regards

**Comment by trinhtranduc** [comment]
I’m having the same issue on macOS Sequoia 15.2. After a successful installation, I don’t see the **PDFWriter** printer in **Printers & Scanners** and I also don’t see any **PDFWriter** files in **/private/var/spool/.**

When I click on the **PDFWriter** alias, it shows an error.

![Screenshot 2024-11-28 at 16 01 53](https://github.com/user-attachments/assets/ffc0854c-cc7e-4164-94c3-35076a8fa19e)

**Comment by rodyager** [comment]
That is the correct behaviour. The executable you are trying to run is designed to be called by the CUPs printing system as it executes the print workflow. It does the job of taking pdf formatted print data drawn by the application that is attempting to print and writing it to a file (instead of sending it to a physical printer)

If you do run it as root it will merely respond with message that indicates the type of data it expects the printing system to give it and identification data that allows Printer Setup to offer it as a Printer.

Rod


Sent from my iPhone

On 28 Nov 2024, at 8:17 pm, Trinh Tran ***@***.***> wrote:

﻿

Open project then build with pdfwriter target and get errror

ERROR: pdfwriter cannot be called without root privileges!

Do you have any suggestion?

**Comment by rodyager** [comment]
The subfolders  are created when you first print (if they don’t exist already)

You have to add the Printer in Printers and scanners. Installing merely offers this as a choice (instead the same way plugging in a physical printer does)

Rod


Sent from my iPhone

On 28 Nov 2024, at 8:02 pm, Trinh Tran ***@***.***> wrote:

﻿

I’m having the same issue on macOS Sequoia 15.2. After a successful installation, I don’t see the PDFWriter printer in Printers & Scanners and I also don’t see any PDFWriter files in /private/var/spool/.

When I click on the PDFWriter alias, it shows an error.

Screenshot.2024-11-28.at.16.01.53.png (view on web)<https://github.com/user-attachments/assets/ffc0854c-cc7e-4164-94c3-35076a8fa19e>

**Comment by trinhtranduc** [comment]
Dear Rod, Thank you for your quick response. However, I don’t fully understand your reply. My primary issue is that I cannot see the printer name in the Printers and Scanners section after successfully completing the installation on macOS 15.2.?

Do you have any suggestions to fix this issue?

**Comment by rodyager** [comment]
I’m running macOS 15.1

Here is the System Setting Printers and scanners before installation.[Screenshot 2024-11-29 at 1.22.58 pm.png]
And here is what happens immediately after running the installer

[Screenshot 2024-11-29 at 1.24.28 pm.png]

The installer is available at

<https://github.com/rodyager/RWTS-PDFwriter/releases/download/v3.1a/RWTS-PDFwriter.pkg>
github.com<https://github.com/rodyager/RWTS-PDFwriter/releases/download/v3.1a/RWTS-PDFwriter.pkg>
        [X] <https://github.com/rodyager/RWTS-PDFwriter/releases/download/v3.1a/RWTS-PDFwriter.pkg>



It is possible that macOS 15.2 breaks things, but that would seem rather unlikely in a point release.


Rod Yager
***@***.***




On 29 Nov 2024, at 12:56 pm, Trinh Tran ***@***.***> wrote:



Dear Rod, Thank you for your quick response. However, I don’t fully understand your reply. My primary issue is that I cannot see the printer name in the Printers and Scanners section after successfully completing the installation on macOS 15.2.?

Do you have any suggestions to fix this issue?

**Comment by rodyager** [comment]
If the driver is not appearing in Printers and Scanners, executing the following command in Terminal might fix the issue or provide clues


sudo lpadmin -p PDFwriter -E -v pdfwriter:/ -P /Library/Printers/PPDs/Contents/Resources/RWTS\ PDFwriter.gz -o printer-is-shared=false



(But this is just the relevant command that the installer executes to add the printer).


Rod

**Comment by trinhtranduc** [comment]
Hi Rob,

I try to run your above command and retrieve an error:

`mac@trinhtran ~ % sudo lpadmin -p PDFwriter -E -v pdfwriter:/ -P /Library/Printers/PPDs/Contents/Resources/RWTS\ PDFwriter.gz -o printer-is-shared=false
Password:
lpadmin: Printer drivers are deprecated and will stop working in a future version of CUPS.
lpadmin: Bad device-uri scheme “pdfwriter”.`

Thanks!

**Comment by trinhtranduc** [comment]
Here is installation log

[log.txt](https://github.com/user-attachments/files/17955598/log.txt)

**Comment by rodyager** [comment]
It may be that the long foreshadowed deprecation of  Printer Drivers has arrived. The warning:
Printer drivers are deprecated and will stop working in a future version of CUPS.
has been around for about a decade.

That is consistent with the subsequent error:
lpadmin: Bad device-uri scheme “pdfwriter”.

If that is the case, it is the end of the line for PDFWriter.

Rod


Rod Yager
***@***.***




On 29 Nov 2024, at 5:59 pm, Trinh Tran ***@***.***> wrote:



Hi Rob,

I try to run your above command and retrieve an error:

***@***.*** ~ % sudo lpadmin -p PDFwriter -E -v pdfwriter:/ -P /Library/Printers/PPDs/Contents/Resources/RWTS\ PDFwriter.gz -o printer-is-shared=false Password: lpadmin: Printer drivers are deprecated and will stop working in a future version of CUPS. lpadmin: Bad device-uri scheme “pdfwriter”.

Thanks!

**Comment by rodyager** [comment]
I’m still on 15.1.1. Updating May well break it for me too


Sent from my iPhone

On 29 Nov 2024, at 6:32 pm, Trinh Tran ***@***.***> wrote:

﻿

Thanks, Rog! I’m just wondering why it’s still working on your Mac version. :)

**Comment by rodyager** [comment]
I have just now updated to macOS 15.2 (Beta 4). The version of cups it uses is still 2.3.4 (you can check this using cups-config --version). Given that, it would seem that the relevant part of the error message is the Bad device-cur scheme.  That most likely means that the pdfwriter executable hasn't been installed in the correct location.   Please execute. 
ls -la /usr/libexec/cups/backend/   
You are expecting to see something like this
lrwxr-xr-x   1 root  wheel      42 30 Nov 19:04 pdfwriter -> /Library/Printers/RWTS/PDFwriter/pdfwriter
where only the Nov 19:04 should be different.
In particular check that the permissions and the ownership are as indicated.

**Comment by trinhtranduc** [comment]
Hi Rod,

Thank you very much for your quick response. You're correct that my cup version is still `2.3.4`. When I run `ls -la /usr/libexec/cups/backend/` and I see same result as you mentioned. However I still not see the printer in the list.

<img width="893" alt="Screenshot 2024-12-03 at 14 47 04" src="https://github.com/user-attachments/assets/12a86007-df10-4626-8a02-75fc3bbf2792">


I tried to manually run command below (modifying your command by replacing `pdfwriter` with `file`)

`lpadmin -p PDFwriter -E -v "file:///Library/Printers/RWTS/PDFwriter/pdfwriter" -P /Library/Printers/PPDs/Contents/Resources/RWTS\ PDFwriter.gz -o printer-is-shared=false`

And now I'm able to see the `PDFwriter` in the list but cannot use this printer.
